src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,615 @@
+        } else if (mode == ReadMode::k_primary_preferred) {
+            node["ReadMode"] = "primaryPreferred";
+        } else if (mode == ReadMode::k_secondary) {
+            node["ReadMode"] = "secondary";
+        } else if (mode == ReadMode::k_secondary_preferred) {
+            node["ReadMode"] = "secondaryPreferred";
+        } else if (mode == ReadMode::k_nearest) {
+            node["ReadMode"] = "nearest";
+        }
+        auto maxStaleness = rhs.max_staleness();
+        if (maxStaleness) {
+            node["MaxStalenessSeconds"] = genny::TimeSpec(*maxStaleness).count();
+        }
+        return node;
+    }
+
+    static bool decode(const Node& node, ReadPreference& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (!node["ReadMode"]) {
+            // readPreference must have a read mode specified.
+            return false;
+        }
+        auto readMode = node["ReadMode"].as<std::string>();
+        rhs = mongocxx::read_preference{};
+        if (readMode == "primary") {
+            rhs.mode(ReadMode::k_primary);
+        } else if (readMode == "primaryPreferred") {
+            rhs.mode(ReadMode::k_primary_preferred);
+        } else if (readMode == "secondary") {
+            rhs.mode(ReadMode::k_secondary);
+        } else if (readMode == "secondaryPreferred") {
+            rhs.mode(ReadMode::k_secondary_preferred);
+        } else if (readMode == "nearest") {
+            rhs.mode(ReadMode::k_nearest);
+        } else {
+            return false;
+        }
+        if (node["MaxStalenessSeconds"]) {
+            auto maxStaleness = node["MaxStalenessSeconds"].as<int>();
+            rhs.max_staleness(std::chrono::seconds(maxStaleness));
+        }
+        return true;
+    }
+};
+
+template <>
+struct convert<mongocxx::write_concern> {
+    using WriteConcern = mongocxx::write_concern;
+    static Node encode(const WriteConcern& rhs) {
+        Node node;
+        auto timeout = rhs.timeout();
+        auto timeoutCount = genny::TimeSpec{timeout}.count();
+        node["TimeoutMillis"] = timeoutCount;
+
+        auto journal = rhs.journal();
+        node["Journal"] = journal;
+
+        auto ackLevel = rhs.acknowledge_level();
+        if (ackLevel == WriteConcern::level::k_majority) {
+            node["Level"] = "majority";
+        } else if (ackLevel == WriteConcern::level::k_acknowledged) {
+            auto numNodes = rhs.nodes();
+            if (numNodes) {
+                node["Level"] = *numNodes;
+            }
+        }
+        return node;
+    }
+
+    static bool decode(const Node& node, WriteConcern& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (!node["Level"]) {
+            // writeConcern must specify the write concern level.
+            return false;
+        }
+        auto level = node["Level"].as<std::string>();
+        bool isNum = level.find_first_not_of("0123456789") == std::string::npos;
```

Comment:
[question] is there a reason you chose to check if the value is a number yourself instead of letting yaml-cpp handle it for you?

```cpp
try {
    auto level = node["Level"].as<int>();
    ...
} catch (const BadConversion& e) {
    auto level = node["Level"].as<std::string>();
    ...
}
```

https://stackoverflow.com/questions/19994312/obtain-type-of-value-stored-in-yamlnode-for-yaml-cpp

Reply:
This slipped my mind :(